File: EasyRecyclerView/src/main/java/cc/easyandroid/easyrecyclerview/EasyProgressRecyclerView.java
package cc.easyandroid.easyrecyclerview;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;

import cc.easyandroid.easyrecyclerview.core.IEmptyAdapter;
import cc.easyandroid.easyrecyclerview.core.IProgressHander;
import cc.easyandroid.easyrecyclerview.core.ProgressEmptyView;
import cc.easyandroid.easyrecyclerview.listener.OnEasyProgressClickListener;

/**
 * EasyProgressRecyclerView
 */
public class EasyProgressRecyclerView extends RecyclerView {

    public EasyProgressRecyclerView(Context context) {
        this(context, null);
    }

    public EasyProgressRecyclerView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, R.attr.EasyRecyclerViewStyle);
    }

    public EasyProgressRecyclerView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        ProgressEmptyView progressEmptyView = new ProgressEmptyView(this, attrs, defStyle);
        setProgressHander(progressEmptyView);

    }

    public void setProgressHander(IProgressHander progressHander) {
        mProgressHander = progressHander;
        setEmptyView(progressHander.getView());
    }

    private IProgressHander mProgressHander;

    public void setAdapter(Adapter adapter) {

        Adapter oldAdapter = getAdapter();
        if (oldAdapter != null) {
            oldAdapter.unregisterAdapterDataObserver(emptyObserver);
        }
        if (adapter != null) {
            adapter.registerAdapterDataObserver(emptyObserver);
        }
        super.setAdapter(adapter);
        emptyObserver.onChanged();
    }


    public void showLoadingView() {
        showLoadingView(null);
    }

    public void showEmptyView() {
        showEmptyView(null);
    }

    public void showErrorView() {
        showErrorView(null);
    }

    public void showLoadingView(String message) {
        mProgressHander.showLoadingView(message);
    }

    public void showEmptyView(String message) {
        mProgressHander.showEmptyView(message);
    }

    public void showErrorView(String message) {
        mProgressHander.showErrorView(message);
    }

    public void setOnEasyProgressClickListener(OnEasyProgressClickListener listener) {
        mProgressHander.setOnEasyProgressClickListener(listener);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        // 要添加在window后才能找到parent
        ViewGroup parent = (ViewGroup) getParent();
        if (parent == null) {
            throw new IllegalStateException(getClass().getSimpleName() + " is not attached to parent view.");
        }
        if (emptyView != null) {
            parent.removeView(emptyView);
            parent.addView(emptyView);
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
    }

    private View emptyView;

    public View getEmptyView() {
        return emptyView;
    }

    private AdapterDataObserver emptyObserver = new AdapterDataObserver() {
        public void onItemRangeChanged(int positionStart, int itemCount) {
            updata();
        }

        public void onItemRangeInserted(int positionStart, int itemCount) {
            updata();
        }

        public void onItemRangeRemoved(int positionStart, int itemCount) {
            updata();
        }

        public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
            updata();
        }

        @Override
        public void onChanged() {
            updata();
        }
    };

    void updata() {
        Adapter<?> adapter = getAdapter();
        if (adapter != null) {
            if (adapter instanceof IEmptyAdapter) {
                IEmptyAdapter iEmptyAdapter = (IEmptyAdapter) adapter;
                if (!iEmptyAdapter.isEmpty() && emptyView != null) {
                    emptyView.setVisibility(View.GONE);
                    EasyProgressRecyclerView.this.setVisibility(View.VISIBLE);
                    return;
                }
            }
            if (emptyView != null) {
                emptyView.setVisibility(View.VISIBLE);
                EasyProgressRecyclerView.this.setVisibility(View.GONE);
            }
        }
    }


    void setEmptyView(View emptyView) {
        this.emptyView = emptyView;
    }

}

File: EasyRecyclerView/src/main/java/cc/easyandroid/easyrecyclerview/EasyProgressRecyclerView.kt
package cc.easyandroid.easyrecyclerview

import android.content.Context
import android.util.AttributeSet
import android.view.View
import kotlin.jvm.JvmOverloads
import cc.easyandroid.easyrecyclerview.R
import androidx.recyclerview.widget.RecyclerView
import cc.easyandroid.easyrecyclerview.core.IProgressHander
import androidx.recyclerview.widget.ConcatAdapter
import cc.easyandroid.easyrecyclerview.core.IEmptyAdapter
import cc.easyandroid.easyrecyclerview.listener.OnEasyProgressClickListener
import android.view.ViewGroup
import cc.easyandroid.easyrecyclerview.core.ProgressEmptyView
import java.lang.IllegalStateException

/**
 * EasyProgressRecyclerView
 */
open class EasyProgressRecyclerView @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyle: Int = R.attr.EasyRecyclerViewStyle,
) : RecyclerView(
    context!!, attrs, defStyle) {
    fun setProgressHander(progressHander: IProgressHander) {
        mProgressHander = progressHander
        emptyView = progressHander.view
    }

    private var mProgressHander: IProgressHander? = null
    override fun setAdapter(adapter: Adapter<*>?) {
        val oldAdapter = getAdapter()
        oldAdapter?.let {//先取消注册
            if (it is IEmptyAdapter) {
                if (it.hasObservers()) {
                    it.unregisterAdapterDataObserver(emptyObserver)
                }
            } else if (it is ConcatAdapter) {
                val adapters = it.adapters
                adapters.forEach { childAdapter ->
                    if (childAdapter is IEmptyAdapter) {
                        if (childAdapter.hasObservers()) {
                            childAdapter.unregisterAdapterDataObserver(emptyObserver)
                        }
                    }
                }
            }
        }

        adapter?.let {//重新注册
            if (it is IEmptyAdapter) {
                it.registerAdapterDataObserver(emptyObserver)
            } else if (it is ConcatAdapter) {
                val adapters = it.adapters
                adapters.forEach { childAdapter ->
                    if (childAdapter is IEmptyAdapter) {
                        childAdapter.registerAdapterDataObserver(emptyObserver)
                    }
                }
            }
        }
        super.setAdapter(adapter)
        emptyObserver.onChanged()
    }

    @JvmOverloads
    fun showLoadingView(message: String? = null) {
        mProgressHander!!.showLoadingView(message)
    }

    @JvmOverloads
    fun showEmptyView(message: String? = null) {
        mProgressHander!!.showEmptyView(message)
    }

    @JvmOverloads
    fun showErrorView(message: String? = null) {
        mProgressHander!!.showErrorView(message)
    }

    fun setOnEasyProgressClickListener(listener: OnEasyProgressClickListener?) {
        mProgressHander!!.setOnEasyProgressClickListener(listener)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        // 要添加在window后才能找到parent
        val parent = parent as ViewGroup
            ?: throw IllegalStateException(javaClass.simpleName + " is not attached to parent view.")
        if (emptyView != null) {
            parent.removeView(emptyView)
            parent.addView(emptyView)
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
    }

    var emptyView: View? = null
    var emptyObserver: AdapterDataObserver = object : AdapterDataObserver() {
        override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
            updata()
        }

        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            updata()
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            updata()
        }

        override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
            updata()
        }

        override fun onChanged() {
            updata()
        }
    }


    fun updata() {
        val adapter = adapter
        if (emptyView != null) {
            if (adapter is ConcatAdapter) {
                val adapters = adapter.adapters
                for (childAdapter in adapters) { //遍历子adapter 进行判断
                    if (handleEmptyAdapter(childAdapter)) return
                }
            } else {
                if (handleEmptyAdapter(adapter)) return
            }
        }
    }

    private fun handleEmptyAdapter(ada: Adapter<*>?): Boolean {
        if (ada is IEmptyAdapter) {
            val iEmptyAdapter = ada as IEmptyAdapter
            if (!iEmptyAdapter.isEmpty) {
                emptyView!!.visibility = GONE
                this@EasyProgressRecyclerView.visibility = VISIBLE
            } else {
                emptyView!!.visibility = VISIBLE
                this@EasyProgressRecyclerView.visibility = GONE
            }
            return true
        }
        return false
    }

    init {
        val progressEmptyView = ProgressEmptyView(this, attrs, defStyle)
        setProgressHander(progressEmptyView)
    }
}